From eea9d5c90d114e32fa0e66f357dbe2f7f4077919 Mon Sep 17 00:00:00 2001
From: Jeremy Huddleston Sequoia <jeremyhu@apple.com>
Date: Sun, 4 Jan 2015 13:30:53 -0800
Subject: [PATCH 7/9] libunwind: Fix fallback implementation of
 _dyld_find_unwind_sections for x86_64 and document dependency on a newer
 libmacho

http://www.llvm.org/bugs/show_bug.cgi?id=22270

Signed-off-by: Jeremy Huddleston Sequoia <jeremyhu@apple.com>
(cherry picked from commit 4d644e4bfc1b3b430cb5edec884f1a24a9b8de80)
---
 src/AddressSpace.hpp | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/src/AddressSpace.hpp b/src/AddressSpace.hpp
index 567cbda..2554e11 100644
--- a/src/AddressSpace.hpp
+++ b/src/AddressSpace.hpp
@@ -316,14 +316,20 @@ LocalAddressSpace::getEncodedP(pint_t &addr, pint_t end, uint8_t encoding,
     // In 10.7.0 or later, libSystem.dylib implements this function.
     extern "C" bool _dyld_find_unwind_sections(void *, dyld_unwind_sections *);
   #else
-    // In 10.6.x and earlier, we need to implement this functionality.
+    // In 10.6.x and earlier, we need to implement this functionality. Note
+    // that this requires a newer version of libmacho (from cctools) than is
+    // present in libSystem on 10.6.x (for getsectiondata).
     static inline bool _dyld_find_unwind_sections(void* addr, 
                                                     dyld_unwind_sections* info) {
       // Find mach-o image containing address.
       Dl_info dlinfo;
       if (!dladdr(addr, &dlinfo))
         return false;
-      const mach_header *mh = (const mach_header *)dlinfo.dli_saddr;
+#if __LP64__
+      const struct mach_header_64 *mh = (const struct mach_header_64 *)dlinfo.dli_saddr;
+#else
+      const struct mach_header *mh = (const struct mach_header *)dlinfo.dli_saddr;
+#endif
       
       // Find dwarf unwind section in that image.
       unsigned long size;
@@ -332,7 +338,7 @@ LocalAddressSpace::getEncodedP(pint_t &addr, pint_t end, uint8_t encoding,
         return false;
       
       // Fill in return struct.
-      info->mh = mh;
+      info->mh = (const struct mach_header *)mh;
       info->dwarf_section = p;
       info->dwarf_section_length = size;
       info->compact_unwind_section = 0;
-- 
2.6.2

